### ISSUE TITLE
MATLAB Projects feature creates hundreds of XML files under resources/project directory

### ISSUE THREAD

**Issue by jvcarli** [mathworks/toolboxdesign#24]
Hi,

While working with the MATLAB Projects feature, I noticed that it generated over 500 XML files under `resources/project`. This number of files seems excessive and introduces unnecessary complexity.

I came across a mention of a "single file project" mechanism in [this pull request](https://github.com/mathworks/climatedatastore/pull/16/files) from `climatedatastore` toolbox, which seems like a more efficient alternative. However, I couldn’t find any official documentation explaining this feature or how to use it.

Could you please confirm if this is an officially supported feature? If so, it would be greatly appreciated to have detailed documentation or guidance on enabling and working with single file projects.

Thank you!

**Comment by rpurser47** [comment]
Yes -- the project system is optimized for very large teams working with thousands of files under source control, where the meta data needs to be stored in a distributed way.  If you have a small group of developers, you can use the "single XML file" version of projects.  Unfortunately, this has to be set **before** you create the project.  In the preferences panel for projects, there is a "Project Definition Files" setting that you want to switch to "use a single project file"
<img width="452" alt="image" src="https://github.com/user-attachments/assets/a0037ce6-dfbc-4762-affb-63bff64bb644">
(the "not recommended" note is specifically for projects with lots of collaborators)

Unfortunately, you'll have to delete the project, commit to Git, change the setting, recreate the project, and then recommit (the one XML file) to git.

**Comment by aforsythe** [comment]
There’s a lot of useful insight in this thread. And with 2025+ it looks like we now need to rely on MATLAB projects to drive toolbox packaging as well. What’s frustrating is the way MathWorks implemented projects on disk. The root folder ends up cluttered with a minimal project file that’s basically just a pointer, and then all the real content lives under a resources/ folder with its own structure of hundreds of XML files (or a single, deeply nested Projects.xml).

This feels pretty at odds to Git workflows. It seems like MATLAB could have represented this as a single root-level project manifest instead of scattering configuration across generated XML. I know Python has different configuration issues, but something closer to uv’s pyproject.toml would cover the vast majority of project setups.

Honestly, the whole workflow feels very 1990s. Configuration that’s technically readable, but not safely editable, spread across a bunch of generated XML artifacts. It also feels backwards that the GUI is treated as the source of truth. The configuration should be the canonical representation of the project, and the GUI should just be a front-end for viewing/editing it. It's not very git friendly since resolving conflicts or making small intentional changes often requires hand-editing.  A stable text-first project manifest would make projects dramatically more version-control friendly.

**Comment by barnesdavid** [comment]
Hi @aforsythe! These are great points, and I am glad to see that you are thinking about the direction of Projects in this way. While there is not anything on our website as it has not been released yet, we are migrating towards a TOML based format (matlab.toml) soon. There are also some additional features coming to Projects to support a fuller package management experience. You can check out this thread for some additional information:
https://github.com/mathworks/MATLAB-extension-for-vscode/issues/282#issuecomment-3442967040
